src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,669 @@
+            {
+                // This is the scenario when the session is created out of Roslyn, in some other provider, e.g. in Debugger.
+                // For now, the default hasSuggestedItemOptions is false. We can discuss if the opposite is required.
+                hasSuggestedItemOptions = false;
+            }
+
+            var filterText = session.ApplicableToSpan.GetText(data.Snapshot);
+            var reason = data.Trigger.Reason;
+
+            // Check if the user is typing a number. If so, only proceed if it's a number
+            // directly after a <dot>. That's because it is actually reasonable for completion
+            // to be brought up after a <dot> and for the user to want to filter completion
+            // items based on a number that exists in the name of the item. However, when
+            // we are not after a dot (i.e. we're being brought up after <space> is typed)
+            // then we don't want to filter things. Consider the user writing:
+            //
+            //      dim i =<space>
+            //
+            // We'll bring up the completion list here (as VB has completion on <space>).
+            // If the user then types '3', we don't want to match against Int32.
+            if (filterText.Length > 0 && char.IsNumber(filterText[0]))
+            {
+                if (!IsAfterDot(data.Snapshot, session.ApplicableToSpan))
+                {
+                    session.Dismiss();
+                    return new AsyncCompletionData.FilteredCompletionModel(ImmutableArray<AsyncCompletionData.CompletionItemWithHighlight>.Empty, 0);
+                }
+            }
+
+            // We need to filter if a non-empty strict subset of filters are selected
+            var selectedFilters = data.SelectedFilters.Where(f => f.IsSelected).Select(f => f.Filter).ToImmutableArray();
+            var needToFilter = selectedFilters.Length > 0 && selectedFilters.Length < data.SelectedFilters.Length;
+
+            var initialListOfItemsToBeIncluded = new List<FilterResult>();
+            foreach (var item in data.InitialSortedList)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (needToFilter && ShouldBeFilteredOutOfCompletionList(item, selectedFilters))
+                {
+                    continue;
+                }
+
+                if (MatchesFilterText(item.FilterText, item.DisplayText, GetMatchPriority(item), filterText, reason))
+                {
+                    initialListOfItemsToBeIncluded.Add(new FilterResult(item, filterText, matchedFilterText: true));
+                }
+                else
+                {
+                    if (reason == AsyncCompletionData.CompletionTriggerReason.Deletion ||
+                        reason == AsyncCompletionData.CompletionTriggerReason.Backspace ||
+                        reason == AsyncCompletionData.CompletionTriggerReason.Invoke ||
+                        filterText.Length <= 1)
```

Comment:
Gotcha.  IS there a reason we borrowed this code, instead of just moving that code ot a common location where both implementation can call into it?  The approach of duplicating logic doesn't seem like a good idea.

Reply:
The only reason to borrow is difference in signatures of the original and the current methods. The original one (FilterModelInBackgroundWorker) accepts Model and returns Model. The new one accepts IAsyncCompletionSession + AsyncCompletionSessionDataSnapshot and returns AsyncCompletionData.FilteredCompletionModel. If not borrowing, we have to implement conversions in two directions + resolving any non-static dependencies with FilterModelInBackgroundWorker. Later, when we will migrate to the new completion, we will have to remove the original model and all conversions. This is my concern.

If you see an efficient solution for this, I'll be happy to consider.

---
In reply to: [226161319](https://github.com/dotnet/roslyn/pull/29016#discussion_r226161319) [](ancestors = 226161319)

Reply:
swap out a front end system, while keepign the majority of the backend the same.  Then, when confident it's all working well, just move the backend entirely forward.  This is what we did for things like FindRefs when we moved to the new UI.  We kept the backend 100% the same.  We just shimmed from the old data to the new data the new UI/controller expected.

Reply:
Thank you, @CyrusNajmabadi! I considered this option. It looks good from the safety perspective. However, it seems that arranging these conversion and modifying the original code to support calls from the new completion look too expensive. So, I'd prefer not doing this now.

---
In reply to: [226435455](https://github.com/dotnet/roslyn/pull/29016#discussion_r226435455) [](ancestors = 226435455)

Reply:
Yes, it is a nice idea. However, the conversion seems to be expensive. So, it seems we cannot make the whole reuse. We have to borrow sometimes.

---
In reply to: [226435873](https://github.com/dotnet/roslyn/pull/29016#discussion_r226435873) [](ancestors = 226435873)